sympify: Made it so that basic numpy datatypes can be cast as Float

bmgxyz opened the pull request:
Factored out changes made in #11666 into a separate function called
convert_numpy_types() in sympify.py. Also called that function both in
sympify.py and in Float() when the input is a numpy type. If the input is a
numpy type that can't be cast as a Float, it passes through to the rest of
sympify. Fixes #13275.

Previously, the following would fail:

```
>>> Float(np.float128(5))
>>> Float(np.float32(5))
```

but now they don't.

<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->

bmgxyz committed:
sympify: Made it so that basic numpy datatypes can be cast as Float

Factored out changes made in #11666 into a separate function called
convert_numpy_types() in sympify.py. Also called that function both in
sympify.py and in Float() when the input is a numpy type. If the input is a
numpy type that can't be cast as a Float, it passes through to the rest of
sympify. Fixes #13275.

Previously, the following would fail:

>>> Float(np.float128(5))
>>> Float(np.float32(5))

but now they don't.

cbm755 commented on sympy/core/numbers.py:
```diff
@@ -967,6 +967,8 @@ def __new__(cls, num, dps=None, prec=None, precision=None):
         elif num is S.NegativeInfinity:
             num = '-inf'
+        elif type(num).__module__ == 'numpy': # support for numpy datatypes
+            num = convert_numpy_types(num)
```
Here you're converting the numpy obj to a Sympy object and letting the rest of the `Float.__new__` deal with the other details.  This looks like the right idea if it works.

cbm755 replied:
Hmmm, what would happen if we just to use `num = sympify(num)` here?  No other changes needed...?

bmgxyz replied:
I initially thought that ```num = sympify(num)``` would cause an endless recursion, but I see now that that's not the case. In fact, after making that change and running it through @Trave11er's tests, it seems to work fine. Incorporated into next commit.

cbm755 commented on sympy/core/sympify.py:
```diff
@@ -51,6 +51,45 @@ class CantSympify(object):
+    >>> x = sp.symbols('x')
+    >>> print(x * np.int64(1))
+    x
+    """
```
I don't think this last docstring adds anything useful.

bmgxyz replied:
Removed.

cbm755 commented on sympy/core/sympify.py:
```diff
@@ -51,6 +51,45 @@ class CantSympify(object):
+    ========
+
+    >>> import sympy as sp
+    >>> import numpy as np
```
This means all doctests will pull in numpy.  If I understand correctly, this is to be avoided.

This function is not "public facing" so maybe it doesn't need these (otherwise nice) doctests.

(I could be wrong about this).

bmgxyz replied:
I've changed this to only import the functions that are used in the doctests. (Is that even necessary? I don't think I fully understand doctests.)

cbm755 commented on sympy/core/sympify.py:
```diff
@@ -260,20 +299,7 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
-                except NotImplementedError:
-                    raise SympifyError('Translation for numpy float : %s '
-                                       'is not implemented' % a)
+        if np.isscalar(a): return convert_numpy_types(a)
```
use two lines here.

bmgxyz replied:
Done.

cbm755 commented on sympy/core/sympify.py:
```diff
@@ -260,20 +299,7 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
     # Support for basic numpy datatypes
     if type(a).__module__ == 'numpy':
         import numpy as np
```
Can this line be removed too?

bmgxyz replied:
I'm not really sure. I think the first time I factored the function out it yelled at me about not knowing what ```numpy``` was. Maybe there's a better way to do this?

cbm755 reviewed (commented):
I like this approach.  Not 100% sure it will work, for that we need tests.

bmgxyz committed:
sympify: Make several changes relating to functionality, doctests, and style

Mainly change Float() constructor so that numpy types get sympified before
being passed on to the rest of the constructor. This avoids a separate
function call. Also, update some doctests and code style to be
smarter/prettier.

bmgxyz commented:
Note that I incorporated @Trave11er's tests into my code locally, but as a courtesy I didn't push them with my most recent commit. They were helpful, though, and my code passed them with one critical exception.

The last three tests (which deal with complex numbers) failed no matter what I tried. That seems to be because ```numpy.complex()``` returns type ```complex```, not a ```numpy``` internal type, so the check in ```sympify.py``` wasn't catching it. No big deal; I just removed those tests locally because (I think) they're not relevant to what we're shooting for (pushing basic ```numpy``` types through ```Float()```)

I am ready to move forward with "rebasing" (however that works) once @Trave11er and @cbm755 are satisfied.

bmgxyz commented:
Oh, one other thing. Now that ```Float.__new__``` doesn't use ```convert_numpy_types()``` anymore, the only place where that function gets used is in ```sympify.py```. Should I un-refactor that function back into ```sympify()``` where it was initially since it's only used once?

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -967,6 +967,8 @@ def __new__(cls, num, dps=None, prec=None, precision=None):
         elif num is S.NegativeInfinity:
             num = '-inf'
+        elif type(num).__module__ == 'numpy': # support for numpy datatypes
+            num = sympify(num)
```
Since you've factored out the code into a function you might as well call `convert_numpy_types` here instead of `sympify`.

asmeurer commented on sympy/core/sympify.py:
```diff
@@ -51,6 +51,41 @@ class CantSympify(object):
     pass
 
 
+def convert_numpy_types(a):
```
I would make this private (start the name with an underscore `_`).

asmeurer replied:
This needs `@doctest_depends_on` since the doctest depends on NumPy (search the codebase for examples on how to use it).

bmgxyz replied:
I tried making ```convert_numpy_types()``` private, but that doesn't work because it gets imported and used by ```numbers.py``` per the change above. I'll leave it public for now (unless I'm missing something).

bmgxyz replied:
Also, I'm having some trouble with ```@doctest_depends_on```. I searched the codebase and inferred that I need to add ```from sympy.utilities.decorator import doctest_depends_on``` to ```sympify.py``` and a ```@doctest_depends_on``` statement later on, but I get an error whenever I add the import to ```sympify.py```. Specifically:

```>>> import sympy
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/__init__.py", line 58, in <module>
    from .core import *
  File "sympy/core/__init__.py", line 4, in <module>
    from .sympify import sympify, SympifyError
  File "sympy/core/sympify.py", line 11, in <module>
    from sympy.utilities.decorator import doctest_depends_on
  File "sympy/utilities/__init__.py", line 4, in <module>
    from .iterables import (flatten, group, take, subsets,
  File "sympy/utilities/iterables.py", line 11, in <module>
    from sympy.core import Basic
ImportError: cannot import name Basic```

Any ideas?

asmeurer commented on sympy/core/sympify.py:
```diff
@@ -51,6 +51,41 @@ class CantSympify(object):
+    >>> from sympy import Float
+    >>> from numpy import int64, float128, float32
+
+    >>> Float(np.int64(5))
```
The doctest should call the function itself, for clarity.

asmeurer commented on sympy/core/sympify.py:
```diff
@@ -51,6 +51,41 @@ class CantSympify(object):
+    >>> Float(np.int64(5))
+    5.00000000000000
+
+    >>> Float(np.float128(5))
```
Is float128 available on all platforms? If not, we should remove this from the doctest so it doesn't fail.

asmeurer replied:
And also I'm not sure even if it is always available if it's always the same precision.

bmgxyz replied:
Okay, for now I'll just remove it.

Trave11er replied:
Can also make it conditional on the platform (see below) but probably easier to remove altogether.
import sys
is_64bits = sys.maxsize > 2**32

asmeurer commented:
Can you add comments by the `type().__name__ == 'numpy'` checks that it's done that way to explicitly avoid importing NumPy.

asmeurer commented:
This needs tests obviously, but I guess the plan is to take them from https://github.com/sympy/sympy/pull/13401.

bmgxyz commented:
@asmeurer Yes, the tests will likely come from there. (In fact I've been using them locally.)

I've implemented many of the changes you suggested, but there are lingering issues. (Mainly with ```@doctest_depends_on```; see above.)

asmeurer commented:
Oh yeah I was afraid the doctest decorator might create circular import issues. I'd have to think of a solution.

asmeurer commented:
I guess we'll have to do this manually https://github.com/sympy/sympy/blob/master/sympy/utilities/decorator.py#L138

bmgxyz commented:
I don't understand what you mean. Do you want me to re-implement ```doctest_depends_on()``` in ```sympify.py```? If so, how? :confused: 

I'm also still not sure why we can't just import stuff at the top of doctests. What am I missing?

asmeurer commented:
Yeah you'll have to add `convert_numpy_types._doctest_depends_on = {'modules': ['numpy']}` after the definition. 

You can't just import it because the utilities submodule imports the core submodule, meaning that if the core module tries to import something from utilities, it creates a circular import (it would recurse infinitely, but the interpreter detects such things and fails). Usually you fix this by moving the imports inside the function so they don't get run at the module level, but that doesn't work for decorators. 

There are also more aggressive solutions possible like removing the core imports from utilities or moving the doctest_depends_on definition to the core, but just doing what it does manually is the simplest (especially since it's just one line).

bmgxyz commented:
That makes a lot of sense; thanks for explaining it to me.

New problem: the manual decorator seems to work fine, but because it's outside the function definition, it doesn't get called when ```convert_numpy_types``` is imported. I messed around with different import statements in the doctest, but I wasn't able to get both (a) ```convert_numpy_types``` to import and (b) the decorator to fire in ```sympify.py``` at the same time. The intricacies of namespaces are eluding me, I suppose. Thoughts?

On the plus side, I'm pretty sure everything else you mentioned above is implemented, and it's passing ```sympy/core``` tests locally (including the new ones in ```test_numbers.py```).

cbm755 commented:
Sorry to be a killjoy, but does this function need doctests?  Its not intended for public usage...

asmeurer commented:
It can be helpful to put doctests in a private function, but if it becomes too complicated to make them work they can be removed.

asmeurer commented:
Can you push up what you tried? It should work to just put that line below the function definition.

bmgxyz committed:
[WIP] sympify: Make some minor changes to convert_numpy_types()

Made several minor adjustments to ```convert_numpy_types``` and its doctests.
Because ```numpy``` is required for the doctests, a manual decorator was
added below ```convert_numpy_types```. This seems to not be working because
it isn't getting called when ```convert_numpy_types``` is imported. Further
work is needed.

This commit fails ```bin/doctest sympy/core/sympify.py``` with the following
error:

```
File "sympy/core/sympify.py", line 61, in
    sympy.core.sympify.convert_numpy_types
Failed example:
    convert_numpy_types(numpy.int64(5))
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest sympy.core.sympify.convert_numpy_types[1]>", line 1, in
        <module>
        convert_numpy_types(numpy.int64(5))
    NameError: name 'numpy' is not defined
**********************************************************************
File "sympy/core/sympify.py", line 64, in
    sympy.core.sympify.convert_numpy_types
Failed example:
    convert_numpy_types(numpy.float32(5))
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest sympy.core.sympify.convert_numpy_types[2]>", line 1, in
        <module>
        convert_numpy_types(numpy.float32(5))
    NameError: name 'numpy' is not defined
```

bmgxyz commented:
@asmeurer Just to clarify, I wasn't able to make the function private because it gets called outside of ```sympify.py``` in ```Float.__new__```.

In general I think I'm missing something obvious here. Maybe one of you can spot the error(s).

cbm755 commented:
I rebased that with the tests, and made it a private function.  @bmgxyz AFAIK this has no meaning in Python beyond convention of an underscore.

I didn't try to keep the doctest stuff.  My preference would be do that as a future PR.

ylemkimon commented:
Closing this PR, since PR #13426 combining this PR is merged.
